include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+	ParameterT operator() const;  // or:  'operator ParameterT() const;'
+
+	// Validity Callbacks (could also use rcl_interfaces::msg::SetParametersResult like in node_parameters.h).
+	// Callbacks are stored in ordered hash map, could also possibly return hash for deleting later on.
+	// Alternatively, return Parameter<ParameterT>& for method chaining
+	using ParameterValidityCallback = std::function<bool(const ParameterT& value, std::string& message)>;
+	Parameter<ParameterT>& warnIf(ParameterValidityCallback warn_if_callback);
+	Parameter<ParameterT>& errorIf(ParameterValidityCallback error_if_callback);  // or rejectIf()/failIf()..
+	// Predefined generic (type-specific?) conditions, i.e. 'NO_OVERRIDE', 'EMPTY', 'NOT_NORMALIZED'
+	enum ParameterCondition
+	{
+		...
+	};
+	Parameter<ParameterT>& warnIf(ParameterCondition condition);
+	Parameter<ParameterT>& errorIf(ParameterCondition condition);
+
+	// Callback for reacting to simple parameter changes (read-locked)
+	// NOTE: this should be called after validity callbacks so that the parameter status can be evaluated
+	using ParameterChangedCallback = std::function<void(const Parameter& parameter)>;
+	void onChanged(ParameterChangedCallback callback);
+
+protected:
+	// Recursively declare parameters and populate child_params_, the actual node parameters (leafs) are
+	// collected and returned so that they can be maintained by NodeConfig. Should call declareChildParam()
+	// internally. NodeParameter is Parameter<rclcpp::Parameter> and needs to be forward declared..
+	virtual std::map<std::string, std::shared_ptr<NodeParameter>> declare();
+
+	// Constructor used for child parameters in the tree that share the same data, so basically all parameters
+	// that are neither root nor leaf.
+	// NOTE: not sure if this is really a good solution or if we should simply copy the data and synchronize
+	// using child/parent callbacks only
+	Parameter(const std::string& name, ParameterT& value, std::shared_mutex parent_mutex);
+
+private:
+	// Initialize child parameter and call declare() function recursively
+	template <ChildParameterT>
+	void declareChildParam(const std::string, std::map<std::string, std::shared_ptr<NodeParameter>>& node_parameters);
+
+	// Internal value protected by mutex_. Alternatively use a struct ParameterValue that manages
+	// locked acces and use that for internal data (this would allow sharing references between child
+	// parameters)
+	ParameterT value_;
+	mutable std::shared_mutex mutex_;
+
+	std::map<std::string, std::shared_ptr<Parameter>> child_params_;
+
+	const std::string name_;
+	std::string description_;
+	std::string additional_constraints_;
+
+	// Store callbacks, use alternative data structurs for ordered hashes
+	std::map<std::string, ParameterValidityCallback> warn_if_callbacks_;
+	std::map<std::string, ParameterValidityCallback> error_if_callbacks_;
+	std::map<std::string, ParameterChangedCallback> on_changed_callbacks_;
```

Comment:
What do all these key strings contain? Shouldn't these all just be lists?

Reply:
Hashes so that we can identify and remove them again? At least for the `on_changed_callbacks_` this would be desirable. But this definitely doesn't have to be a feature for the MVP.

Reply:
A much simpler approach would just use `std::list` then and return the iterator as a handle.
The overhead of maps for this feels unjustified.

Reply:
of course!